File: apps/ios/AceClub/Domain/Entities/MatchIntent.swift
//
//  MatchIntent.swift
//  AceClub
//

import Foundation

// MARK: - Match Intent Status

enum MatchIntentStatus: String, CaseIterable {
    case pending
    case accepted
    case rejected
}

// MARK: - Match Intent Type

enum MatchIntentType: String, CaseIterable {
    case match
    case training

    var displayName: String {
        switch self {
        case .match: return "Match"
        case .training: return "Entraînement"
        }
    }

    var icon: String {
        switch self {
        case .match: return "sportscourt"
        case .training: return "figure.run"
        }
    }
}

// MARK: - Match Intent

struct MatchIntent: Identifiable {
    let id: String
    let userId: String
    let date: Date?
    let time: Date?
    let duration: Int
    let type: MatchIntentType
    let description: String?
    let status: MatchIntentStatus
    let createdAt: Date?
}

// MARK: - List result (cursor pagination)

struct MatchIntentListResult {
    let data: [MatchIntent]
    let nextCursor: String?
    let hasMore: Bool
    let limit: Int
}

// MARK: - User Brief (for discover & requester)

struct UserBrief: Identifiable {
    let id: String
    let name: String
    let email: String
}

// MARK: - Discover item (intent + user)

struct MatchIntentDiscoverItem: Identifiable {
    let intent: MatchIntent
    let user: UserBrief?

    var id: String { intent.id }
}

struct DiscoverListResult {
    let data: [MatchIntentDiscoverItem]
    let nextCursor: String?
    let hasMore: Bool
    let limit: Int
}

// MARK: - Match Request Status

enum MatchRequestStatus: String, CaseIterable {
    case pending
    case accepted
    case rejected
}

// MARK: - Match Request

struct MatchRequest: Identifiable {
    let id: String
    let matchIntentId: String
    let requesterId: String
    let receiverId: String
    let status: MatchRequestStatus
    let createdAt: Date
    let respondedAt: Date?
}

// MARK: - Match Request with details

struct MatchRequestWithDetails: Identifiable {
    let request: MatchRequest
    let matchIntent: MatchIntent?
    let requester: UserBrief?

    var id: String { request.id }
}

// MARK: - Swipe result

struct SwipeResult {
    let matchRequest: MatchRequest?
    let message: String?
}

// MARK: - User Contact (for contact modal after accept)

struct UserContact: Identifiable {
    let id: String
    let name: String
    let image: String?
    let phoneNumber: String?

    var imageURL: URL? {
        guard let image, !image.isEmpty else { return nil }
        return URL(string: image)
    }

    var initials: String {
        let components = name.split(separator: " ")
        if components.count >= 2 {
            let firstInitial = components[0].prefix(1)
            let lastInitial = components[1].prefix(1)
            return "\(firstInitial)\(lastInitial)".uppercased()
        } else if let first = components.first {
            return String(first.prefix(2)).uppercased()
        }
        return "??"
    }
}

// MARK: - Accept request result

struct AcceptMatchRequestResult {
    let match: Match?
    let request: MatchRequest?
    let requester: UserContact?
    let message: String?
}

// MARK: - Reject request result

struct RejectMatchRequestResult {
    let request: MatchRequest?
    let message: String?
}

File: apps/mobile/types/match-intent.ts
import type { CursorPagination } from "./common";

export type MatchIntentStatus = "pending" | "accepted" | "rejected";
export type MatchIntentType = "match" | "training";
export type MatchRequestStatus = "pending" | "accepted" | "rejected";

export interface MatchIntent {
  id: string;
  userId: string;
  date: string | null;
  time: string | null;
  isFlexibleDate: boolean;
  duration: number;
  type: MatchIntentType;
  description?: string | null;
  status: MatchIntentStatus;
  createdAt: string;
}

export interface OrganizationBrief {
  id: string;
  name: string;
  slug: string;
  logo?: string | null;
}

export interface UserBrief {
  id: string;
  name: string;
  image?: string | null;
  level: number;
  skillLevel?: string | null;
  sport?: string | null;
  organization?: OrganizationBrief | null;
}

export interface TeammateBrief {
  slotIndex: number;
  userId: string;
  name: string | null;
  image: string | null;
}

export interface MatchIntentWithUser {
  intent: MatchIntent;
  user?: UserBrief | null;
  distance?: number | null;
  myRequestStatus: MatchRequestStatus | null;
  myRequestSlotIndex: number | null;
  teammates: TeammateBrief[] | null;
}

export interface ListMatchIntentsResponse {
  data: MatchIntent[];
  pagination: CursorPagination;
}

/** Raw item from the discover API (flat structure) */
export interface DiscoverItemRaw {
  id: string;
  userId: string;
  type: string;
  status: string;
  date?: string | null;
  time?: string | null;
  isFlexibleDate: boolean;
  duration?: number | null;
  description?: string | null;
  createdAt: string;
  distance?: number | null;
  user?: UserBrief | null;
  myRequestStatus: MatchRequestStatus | null;
  myRequestSlotIndex: number | null;
  teammates: TeammateBrief[] | null;
}

/** Raw API response from /match-intents/discover */
export interface DiscoverResponseRaw {
  data: DiscoverItemRaw[];
  pagination: CursorPagination;
  isDiscoveryRestricted: boolean;
}

/** Mapped response used by the UI */
export interface DiscoverResponse {
  data: MatchIntentWithUser[];
  pagination: CursorPagination;
  isDiscoveryRestricted: boolean;
}

export interface CreateMatchIntentRequest {
  date?: string;
  time?: string;
  isFlexibleDate?: boolean;
  duration: number;
  type?: MatchIntentType;
  description?: string;
  teammateUserIds?: string[];
}

export interface CreateRequestRequest {
  slotIndex?: number;
}

export interface CreateRequestResponse {
  request: MatchRequest;
  conversationId: string;
  message: unknown;
}

export interface MatchRequest {
  id: string;
  matchIntentId: string;
  requesterId: string;
  receiverId: string;
  slotIndex: number | null;
  status: MatchRequestStatus;
  createdAt: string;
  respondedAt?: string | null;
}

export interface MatchRequestWithDetails extends MatchRequest {
  matchIntent?: MatchIntent | null;
  requester?: UserBrief | null;
}

export interface UserContact {
  id: string;
  name: string;
  image?: string | null;
}

export interface AcceptMatchRequestResponse {
  match?: { id: string } | null;
  request?: MatchRequest | null;
  requester?: UserContact | null;
  conversationId?: string | null;
  teamComplete?: boolean;
  message?: string | null;
}

export interface RejectMatchRequestResponse {
  request?: MatchRequest | null;
  message?: string | null;
}
